src/musig.rs

```diff
@@ -0,0 +1,1158 @@
+                unreachable!("Serialization cannot fail")
+            } else {
+                data.assume_init()
+            }
+        }
+    }
+
+    /// Deserialize a PartialSignature from bytes.
+    ///
+    /// # Errors:
+    ///
+    /// - MalformedArg: If the signature [`PartialSignature`] is out of curve order
+    pub fn from_byte_array(data: &[u8; ffi::MUSIG_PART_SIG_SERIALIZED_LEN]) -> Result<Self, ParseError> {
+        let mut partial_sig = MaybeUninit::<ffi::MusigPartialSignature>::uninit();
+        unsafe {
+            if ffi::secp256k1_musig_partial_sig_parse(
+                ffi::secp256k1_context_no_precomp,
+                partial_sig.as_mut_ptr(),
+                data.as_ptr(),
+            ) == 0
+            {
+                Err(ParseError::MalformedArg)
+            } else {
+                Ok(PartialSignature(partial_sig.assume_init()))
+            }
+        }
+    }
+
+    /// Get a const pointer to the inner PartialSignature
+    pub fn as_ptr(&self) -> *const ffi::MusigPartialSignature { &self.0 }
+
+    /// Get a mut pointer to the inner PartialSignature
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::MusigPartialSignature { &mut self.0 }
+}
+
+impl KeyAggCache {
+    /// Creates a new [`KeyAggCache`] by supplying a list of PublicKeys used in the session.
+    ///
+    /// Computes a combined public key and the hash of the given public keys.
+    ///
+    /// Different orders of `pubkeys` result in different `agg_pk`s.
+    /// The pubkeys can be sorted lexicographically before combining with which
+    /// ensures the same resulting `agg_pk` for the same multiset of pubkeys.
+    /// This is useful to do before aggregating pubkeys, such that the order of pubkeys
+    /// does not affect the combined public key.
+    ///
+    /// # Returns
+    ///
+    ///  A [`KeyAggCache`] the can be used [`KeyAggCache::nonce_gen`] and [`Session::new`].
+    ///
+    /// # Args:
+    ///
+    /// * `secp` - Secp256k1 context object initialized for verification
+    /// * `pubkeys` - Input array of public keys to combine. The order is important; a
+    ///   different order will result in a different combined public key
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// # # [cfg(any(test, feature = "rand-std"))] {
+    /// # use secp256k1::rand::{thread_rng, RngCore};
+    /// # use secp256k1::{KeyAggCache, Secp256k1, SecretKey, Keypair, PublicKey};
+    /// # let secp = Secp256k1::new();
+    /// # let sk1 = SecretKey::new(&mut thread_rng());
+    /// # let pub_key1 = PublicKey::from_secret_key(&secp, &sk1);
+    /// # let sk2 = SecretKey::new(&mut thread_rng());
+    /// # let pub_key2 = PublicKey::from_secret_key(&secp, &sk2);
+    /// #
+    /// let key_agg_cache = KeyAggCache::new(&secp, &[pub_key1, pub_key2]);
+    /// let _agg_pk = key_agg_cache.agg_pk();
+    /// # }
+    /// ```
+    pub fn new<C: Verification>(secp: &Secp256k1<C>, pubkeys: &[&PublicKey]) -> Self {
```

Comment:
In dd26c3067044a89c503a32079c24227d6e1afca5:

What if you pass an empty array of pubkeys? I'll play with this. Probably I'll just add an explict panic, though given that we have the same issue with partial sigs, I'm tempted to add an error variant for this, or maybe a nonemptyarray newtype. Will address in followup, at least to add a panic (I suspect the existing code will do weird potentially-UB things) but I might wind up just filing an issue.

Reply:
For your curiosity, what happens is that we try to pass a bad pointer to the C code, but our alignment trickery in secp256k1-sys catches it and panics (though I doubt the panic is very reliable). So on my system it's not actually UB, but it's pretty close.